### PR TITLE
End-to-End Build Workflow

### PR DESCRIPTION
**Pull request by okankoAMZ**
# Description of the issue
Allows us to build and test with operator for full end-to-end testing

# Description of changes
Added new workflow to build agent and operator together

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
[Build Test Artifacts](https://github.com/aws/amazon-cloudwatch-agent/actions/workflows/build-test-artifacts.yml)
https://github.com/aws/amazon-cloudwatch-agent/actions/runs/11748578760
Example of workflow call working as intended for BuildArtifacts: https://github.com/aws/amazon-cloudwatch-agent/actions/runs/12321954899

# Requirements
_Before commit the code, please do the following steps._
1. Run `make fmt` and `make fmt-sh`
2. Run `make lint`

### CODE REVIEW

File: .github/workflows/e2e-build.yml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            cancel-in-progress: true
+        secrets: inherit
+        with:
+            tag: ${{needs.GetLatestOperatorCommitSHA.outputs.operator_commit_sha}}
+            target-branch: ${{needs.GetLatestOperatorCommitSHA.outputs.operator_commit_sha}}
+            repository:  ${{needs.GetLatestOperatorCommitSHA.outputs.operator_repo_name}}
+            test-image-before-upload: false            
+
+
+
```
Comment (musa-asad) [inline .github/workflows/e2e-build.yml:69]: nit:
```suggestion
```

File: .github/workflows/e2e-build.yml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        permissions:
+          id-token: write
+          contents: read
+        with:
+            test-image-before-upload: false            
+    BuildOperator:
+        needs: [GetLatestOperatorCommitSHA]
+        uses: aws/amazon-cloudwatch-agent-operator/.github/workflows/build-and-upload.yml@e2e-testing
+        concurrency:
+            group: ${{ github.workflow }}-operator-${{ inputs.operator-branch}}
```
Comment (musa-asad) [inline .github/workflows/e2e-build.yml:59]: nit:
```suggestion
            group: ${{ github.workflow }}-operator-${{ inputs.operator-branch }}
```

File: .github/workflows/e2e-build.yml [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            cancel-in-progress: true
+        secrets: inherit
+        with:
+            tag: ${{needs.GetLatestOperatorCommitSHA.outputs.operator_commit_sha}}
+            target-branch: ${{needs.GetLatestOperatorCommitSHA.outputs.operator_commit_sha}}
+            repository:  ${{needs.GetLatestOperatorCommitSHA.outputs.operator_repo_name}}
```
Comment (musa-asad) [inline .github/workflows/e2e-build.yml:65]: nit:
```suggestion
            tag: ${{ needs.GetLatestOperatorCommitSHA.outputs.operator_commit_sha }}
            target-branch: ${{ needs.GetLatestOperatorCommitSHA.outputs.operator_commit_sha }}
            repository:  ${{ needs.GetLatestOperatorCommitSHA.outputs.operator_repo_name }}
```

File: .github/workflows/e2e-build.yml [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+          - name: Get latest commit SHA
+            id: get_latest_sha
+            run: |
+                    cd operator-repo
+                    latest_sha=$(git rev-parse HEAD)
+                    echo "::set-output name=operator_sha::$latest_sha"
+    BuildAgent:
+        uses: ./.github/workflows/build-test-artifacts.yml
+        concurrency:
+            group: "Build-Test-Artifacts-${{github.ref_name}}"
```
Comment (musa-asad) [inline .github/workflows/e2e-build.yml:47]: nit:
```suggestion
            group: "Build-Test-Artifacts-${{ github.ref_name }}"
```

File: .github/workflows/e2e-build.yml [hunk 5]
```diff
@@ -0,0 +1,69 @@
+
+
```
Comment (musa-asad) [inline .github/workflows/e2e-build.yml:2]: nit:
```suggestion
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT
```

File: .github/workflows/e2e-build.yml [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        runs-on: ubuntu-latest
+        outputs:
+            operator_commit_sha: ${{steps.get_latest_sha.outputs.operator_sha}}
+            operator_repo_name: ${{env.OPERATOR_GITHUB_REPO_NAME}} 
+        steps:
+          - name: Checkout the target repo
+            uses: actions/checkout@v3
+            with:
+                repository: ${{env.OPERATOR_GITHUB_REPO_NAME}} 
+                ref: ${{inputs.operator-branch}}
```
Comment (musa-asad) [inline .github/workflows/e2e-build.yml:35]: nit:
```suggestion
            operator_commit_sha: ${{ steps.get_latest_sha.outputs.operator_sha }}
            operator_repo_name: ${{ env.OPERATOR_GITHUB_REPO_NAME }} 
        steps:
          - name: Checkout the target repo
            uses: actions/checkout@v3
            with:
                repository: ${{ env.OPERATOR_GITHUB_REPO_NAME }} 
                ref: ${{ inputs.operator-branch }}    

```

File: .github/workflows/e2e-build.yml [hunk 7]
```diff
@@ -0,0 +1,69 @@
+
+
+
+name: Build End-to-End Test Artifacts
+env:
+    OPERATOR_GITHUB_REPO_NAME: "aws/amazon-cloudwatch-agent-operator"
+    ECR_TARGET_ALLOCATOR_STAGING_REPO: ${{ vars.ECR_TARGET_ALLOCATOR_STAGING_REPO}}
+    ECR_OPERATOR_RELEASE_IMAGE: ${{ vars.ECR_TARGET_ALLOCATOR_TEST_OPERATOR_REPO}}
```
Comment (musa-asad) [inline .github/workflows/e2e-build.yml:8]: nit:
```suggestion
    ECR_TARGET_ALLOCATOR_STAGING_REPO: ${{ vars.ECR_TARGET_ALLOCATOR_STAGING_REPO }}
    ECR_OPERATOR_RELEASE_IMAGE: ${{ vars.ECR_TARGET_ALLOCATOR_TEST_OPERATOR_REPO }}
```

**Review by musa-asad — DISMISSED** [review]

File: .github/workflows/e2e-build.yml [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    workflow_dispatch:
+        inputs:
+            operator-branch:
+                required: true
+                type: string
+                default: 'main'
+    workflow_call:
+        inputs:
+            operator-branch:
+                required: true
```
Comment (musa-asad) [inline .github/workflows/e2e-build.yml:20]: If we have a default, does it need to be required?

File: .github/workflows/e2e-build.yml [hunk 9]
```diff
@@ -0,0 +1,69 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT
+
+name: Build End-to-End Test Artifacts
+env:
+    OPERATOR_GITHUB_REPO_NAME: "aws/amazon-cloudwatch-agent-operator"
+    ECR_TARGET_ALLOCATOR_STAGING_REPO: ${{ vars.ECR_TARGET_ALLOCATOR_STAGING_REPO}}
+    ECR_OPERATOR_RELEASE_IMAGE: ${{ vars.ECR_TARGET_ALLOCATOR_TEST_OPERATOR_REPO}}
```
Comment (musa-asad) [inline .github/workflows/e2e-build.yml:8]: These aren't being used?

File: .github/workflows/e2e-build.yml [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    BuildOperator:
+        needs: [GetLatestOperatorCommitSHA]
+        uses: aws/amazon-cloudwatch-agent-operator/.github/workflows/build-and-upload.yml@e2e-testing
+        concurrency:
+            group: ${{ github.workflow }}-operator-${{ inputs.operator-branch}}
+            cancel-in-progress: true
+        secrets: inherit
+        with:
+            tag: ${{needs.GetLatestOperatorCommitSHA.outputs.operator_commit_sha}}
+            target-branch: ${{needs.GetLatestOperatorCommitSHA.outputs.operator_commit_sha}}
```
Comment (musa-asad) [inline .github/workflows/e2e-build.yml:64]:
```suggestion
            target-sha: ${{needs.GetLatestOperatorCommitSHA.outputs.operator_commit_sha}}
```

File: .github/workflows/e2e-build.yml [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+            id: get_latest_sha
+            run: |
+                    cd operator-repo
+                    latest_sha=$(git rev-parse HEAD)
+                    echo "::set-output name=operator_sha::$latest_sha"
+    BuildAgent:
+        uses: ./.github/workflows/build-test-artifacts.yml
+        concurrency:
+            group: "Build-Test-Artifacts-${{github.ref_name}}"
+            cancel-in-progress: true
```
Comment (musa-asad) [inline .github/workflows/e2e-build.yml:48]: Do we need this concurrency? The workflow it calls already has one.

File: .github/workflows/build-test-artifacts.yml [hunk 12]
```diff
@@ -16,6 +16,12 @@ on:
       - '!.github/workflows/integration-test.yml'
       - '!.github/workflows/application-signals-e2e-test.yml'
   workflow_dispatch:
+  workflow_call:
+    inputs:
+      test-image-before-upload:
+        description: "Run Test on the new container image"
+        default: true
+        type: boolean
```
Comment (musa-asad) [inline .github/workflows/build-test-artifacts.yml:24]:
```suggestion
  workflow_dispatch:
    inputs:
      test-image-before-upload:
        description: "Run Test on the new container image"
        default: true
        type: boolean
  workflow_call:
    inputs:
      test-image-before-upload:
        description: "Run Test on the new container image"
        default: true
        type: boolean
```

File: .github/workflows/e2e-build.yml (hunk 9 above)
Comment (musa-asad) [inline .github/workflows/e2e-build.yml:8]:
```suggestion
    OPERATOR_GITHUB_REPO_NAME: "aws/amazon-cloudwatch-agent-operator"
```

File: .github/workflows/e2e-build.yml [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+env:
+    OPERATOR_GITHUB_REPO_NAME: "aws/amazon-cloudwatch-agent-operator"
+    ECR_TARGET_ALLOCATOR_STAGING_REPO: ${{ vars.ECR_TARGET_ALLOCATOR_STAGING_REPO}}
+    ECR_OPERATOR_RELEASE_IMAGE: ${{ vars.ECR_TARGET_ALLOCATOR_TEST_OPERATOR_REPO}}
+
+on:
+    workflow_dispatch:
+        inputs:
+            operator-branch:
+                required: true
```
Comment (musa-asad) [inline .github/workflows/e2e-build.yml:14]:
```suggestion
                required: false
```

File: .github/workflows/e2e-build.yml [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+                type: string
+                default: 'main'
+    workflow_call:
+        inputs:
+            operator-branch:
+                required: true
+                type: string
+                default: 'main'
```
Comment (musa-asad) [inline .github/workflows/e2e-build.yml:22]:
```suggestion
on:
    workflow_dispatch:
        inputs:
            operator-branch:
                required: false
                type: string
                default: 'main'
    workflow_call:
        inputs:
            operator-branch:
                required: false
                type: string
                default: 'main'
```

File: .github/workflows/e2e-build.yml [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+            cancel-in-progress: true
+        secrets: inherit
+        permissions:
+          id-token: write
+          contents: read
+        with:
+            test-image-before-upload: false            
+    BuildOperator:
+        needs: [GetLatestOperatorCommitSHA]
+        uses: aws/amazon-cloudwatch-agent-operator/.github/workflows/build-and-upload.yml@e2e-testing
```
Comment (musa-asad) [inline .github/workflows/e2e-build.yml:57]:
```suggestion
        uses: aws/amazon-cloudwatch-agent-operator/.github/workflows/build-and-upload.yml@main
```

**Review by musa-asad — CHANGES_REQUESTED** [review]
Please make these minor changes, otherwise it's good to go!

File: .github/workflows/build-test-artifacts.yml [hunk 16]
```diff
@@ -87,6 +93,7 @@ jobs:
 
   StartIntegrationTests:
     needs: [ BuildAndUploadPackages, BuildAndUploadITAR, BuildAndUploadCN, BuildDocker ]
+    if: ${{inputs.test-image-before-upload || true }}
```
Comment (musa-asad) [inline .github/workflows/build-test-artifacts.yml:96]:
```suggestion
    if: ${{true || inputs.test-image-before-upload }}
```

File: .github/workflows/build-test-artifacts.yml [hunk 17]
```diff
@@ -96,7 +103,7 @@ jobs:
   StartApplicationSignalsE2ETests:
     needs: [ BuildAndUploadPackages, BuildAndUploadITAR, BuildAndUploadCN, BuildDocker ]
     # Workflow only runs against main
-    if: ${{ contains(github.ref_name, 'main') }}
+    if: ${{ contains(github.ref_name, 'main')  && (inputs.test-image-before-upload || true) }}
```
Comment (musa-asad) [inline .github/workflows/build-test-artifacts.yml:106]:
```suggestion
    if: ${{ contains(github.ref_name, 'main')  && (true || inputs.test-image-before-upload) }}
```

File: .github/workflows/build-test-artifacts.yml [hunk 18]
```diff
@@ -87,6 +93,7 @@ jobs:
 
   StartIntegrationTests:
     needs: [ BuildAndUploadPackages, BuildAndUploadITAR, BuildAndUploadCN, BuildDocker ]
+    if: ${{inputs.test-image-before-upload == null || inputs.test-image-before-upload}}
```
Comment (musa-asad) [inline .github/workflows/build-test-artifacts.yml:96]:
```suggestion
    if: ${{ (github.event_name != 'workflow_call') || inputs.test-image-before-upload }}
```

File: .github/workflows/build-test-artifacts.yml [hunk 19]
```diff
@@ -96,7 +103,7 @@ jobs:
   StartApplicationSignalsE2ETests:
     needs: [ BuildAndUploadPackages, BuildAndUploadITAR, BuildAndUploadCN, BuildDocker ]
     # Workflow only runs against main
-    if: ${{ contains(github.ref_name, 'main') }}
+    if: ${{ contains(github.ref_name, 'main')  && (inputs.test-image-before-upload == null ||inputs.test-image-before-upload) }}
```
Comment (musa-asad) [inline .github/workflows/build-test-artifacts.yml:106]:
```suggestion
    if: ${{ contains(github.ref_name, 'main') && ((github.event_name != 'workflow_call') || inputs.test-image-before-upload) }}
```

**Review by varunch77 — APPROVED** [review]

**Review by TravisStark — APPROVED** [review]

**Pull request merged**